---
title: "Introduction to microCompet"
author: "Zhuyi Wang"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to microCompet}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
library(knitr)
opts_chunk$set(fig.align = "center", 
               out.width = "80%",
               fig.width = 1, fig.height = 0.6,
               dev.args=list(pointsize=10),
               par = TRUE, # needed for setting hook 
               collapse = TRUE, # collapse input & output code in chunks
               warning = FALSE)
knit_hooks$set(par = function(before, options, envir)
  { if(before && options$fig.show != "none") 
       par(family = "sans", mar=c(4.1,4.1,1.1,1.1), mgp=c(3,1,0), tcl=-0.5)
})
set.seed(1) # for exact reproducibility
```

__*Welcome to the world of microbes and microbiota!*__

<img src="../inst/extdata/microbiome.jpg" height="400px" width="600px"/>


```{r, eval=FALSE}
- Introduction
- Available Datasets
- Available Functions
- References
```


## Introduction
`microCompet` is a package for visualizing potential microbial competition for simple sugars as sources of nutrition and energy. It can plot all pathways one microbe has, evaluate each pathway's completeness, and compare across species to score the potential for competition. 

To use this package, the only required input from users is a microbial genome in the format of an __*ANNOTATED* GenBank file__. By annotated, this means "genes" are labeled onto a genome, and the file is not one very very very long single line of ATCG. See provided files in the main directory for example, __*Klebsiella_variicola.gb*__, and __*Lactobacillus_johnsonii.gb*__. (Microbe names are hard to pronounce, but it's harder to resist the temptation to try!).

The first step is to load the library, and following this vignettes do not require any other user input.
```{r, eval=FALSE, message=FALSE, warning=FALSE}
require("microCompet")
# and these are all the dependecies, all available from CRAN
# require("radarchart")
# require("ggraph")
# require("network")
# require("sna")
# require("ggplot2")
# require("igraph")
# require("dplyr")
```


## Available Datasets
__1. EnzymaticReactions__

Each row in this dataset uses 6 attributes to describe an enzymatic reaction.

If a reaction equation is substrate1 ---> product1, catalyzed by enzyme E1. Now we obviously have *Substrate*, *Product*, and *Enzyme*. *Gene* is the name of a gene that encodes enzyme E1, and different microbial species may different genes encoding slightly different versions E1. *Reaction.EC* is a universal naming system for enzymes based on the category of reaction it catalyzed, full version is Enzyme Commission, usually in the form of 2.7.1.55. The last *Sugar* is to describe which sugar this enzyme involves in degrading.

If a reaction is substrate1 ---> product1 + product2, then they are written as two separate lines with one product per line (all other attributes would be the same). This is for easier network presentation.

<br>

__2. EnzymeDistribution__

Each row describes one gene/enzye and its distribution in microbes. The first 4 columns __Gene__, __Reaction.EC__, __Enzyme__, __Sugar__ are the same as described in __*EnzymaticReactions*__. While column 5 to 13 describe 9 genomes, a "0" in a cell suggests this species/genome does not encode the gene in the row, while "1" suggests it's present. These 9 species are named by first letter of genus and first four letters of species, thus Klebsiella variicola is "Kvari". Check their full name by: `colnames(EnzymeDistribution)[5:13]`

<br>

__3. GenomesInfo__

This dataset lists information about genomes included in dataset __*EnzymeDistribution*__. If you want to run this package with your own datasets, this one is not required. However, remember to run __*checkUserED*__ for your dataset.

__Attention for self-defined dataseets__
If you read in a data frame from a tsv, tab file, be careful about empty spaces in the cell. All functions used here are sensitive the spaces, and "ribose" and "ribose " is not considered as the same sugar. You can call the trim() function in excel before saving the dataset.

## Available Functions
__1. checkUserED__

If users decide to use their own dataset ED, this function helps to check whether required column "Gene" are present in the provided __*ED*__(short name for user-provided __*EnzymeDistribution*__), and whether the chosen columns representing genomes of interest contain only 0 and 1. I assume genome columns are continuous from column with index __*firstMicrobe*__ to __*lastMicrobe*__. See built-in __*EnzymeDistribution*__ for example.

<br>

__2. extractCarboGenes__

When user provide a full path to an annotated genome (a .gb or .gbk file), and a list of potentially sugar degradation genes of interest, this function extract available *carboGenes* from the genome. There are other available GenBank-parsing packages, but they usually format the whole file (about 20M for one microbe) into one very large object, which is very slow and provides unnecessary information for this package. 

Function __*extractCarboGenes*__ simply screen over the file, extract all lines in the format "/gene="gene_name" and select ones in the list of genes provided by users. See example with 
```{r extractCarboGenes}
require("microCompet")
# load package datasets
ED <- microCompet::EnzymeDistribution
# Will include all sugar degradation genes for this example
fullEnzymeGeneList <- ED$Gene
genomeFilePath <- system.file("extdata",
                              "Lactobacillus_johnsonii.gb", #You can also try "Klebsiella_variicola.gb"
                              package = "microCompet",
                              mustWork = TRUE)
carboGenes <- extractCarboGenes(genomeFilePath, fullEnzymeGeneList)
carboGenes
```
<br>

__3. constructFullNetwork__

This function offers to visualize the all sugar degradation pathways of a microbial genome. The three arguments are name of a microbe, a full list of sugar degradation enzymes (output from __*extractCarboGenes*__), and a dataset with all reactions of interest (See __*EnzymaticReactions*__ for example and descriptions). A sample can be produced by:
```{r constructFullNetwork, eval = FALSE, message=FALSE, warning=FALSE, out.width="800px"}
require("microCompet")
ER <- microCompet::EnzymaticReactions
ED <- microCompet::EnzymeDistribution
fullEnzymeGeneList <- ED$Gene
genomeFilePath <- system.file("extdata",
                              "Lactobacillus_johnsonii.gb", #You can also try "Klebsiella_variicola.gb"
                              package = "microCompet",
                              mustWork = TRUE)
carboGenes <- extractCarboGenes(genomeFilePath, fullEnzymeGeneList)
fullPathway <- constructFullNetwork("Lactobacillus johnsonii", carboGenes, ER)
fullPathway
```
<img src="../inst/extdata/fullNet.png" height="400px" width="700px"/>

Don't be scared by many legends, they would make sense after explanation :D

Each vertex is an intermediate compound that appears in the pathway, while each edge suggest enzymes/genes that can catalyze this reaction (tail as substrate and head as product).

__Size of a vertex__ implies its importance in the pathway, compounds that involve in more reactions are larger. __Color of an edge__ identifies the sugar degradation pathway it participates in, and __width of an edge__ indicates number of genes encoding an enzyme that can catalyze this reaction.

Enzyme name or reaction EC for each edge is not labeled. Though this sample figure look quite empty, a species within the __Escherichia__ or __Klebsiella__ genus can usually degrade many different sugars and this image might be too busy. In addition, enzyme versions do not contribute very much to the basic understanding sugar degradation capacity of a bacterial species.

Try this function with __*Klebsiella_variicola.gb*__ or other annotated gbk files. (Check NCBI)

<br>

__4. overallSimilarity__

Given a genome of interest (represented by a list of gene names) and an expression profile of some other species (must be offered in a dataframe following the format of __*EnzymeDistribution*__), this function compares the overall similarity of the genome of interest with other species and produce a radar chart. See example
```{r overSimi, message=FALSE, warning=FALSE}
require("microCompet")
genomeName <- "L. johnsonii"
ER <- microCompet::EnzymaticReactions
ED <- microCompet::EnzymeDistribution
fullEnzymeGeneList <- ED$Gene
# No need to rerun extractCarboGenes (lines with #) if you ran the previous example
genomeFilePath <- system.file("extdata",                              #
                              "Lactobacillus_johnsonii.gb",           #
                              package = "microCompet",                #
                              mustWork = TRUE)                        #
carboGenes <- extractCarboGenes(genomeFilePath, fullEnzymeGeneList)   #
firstMicrobe <- 5 # first column in ED representing a genome
lastMicrobe <- 13 # last column in ED representing a genome
overSimi <- overallSimilarity(genomeName, carboGenes, ED, firstMicrobe, lastMicrobe)
overSimi
```

The full score is the total number of enzymes in the provided dataset. Common features is either "both genome express one enzyme" or "neither express it". One common feature adds 1 to the similarity score, and the radar chart visually tells you which species might be the strongest competitor (most similar). To see full name of each species on the corner, check out the description file for __*EnzymeDistribution*__ by `?EnzymeDistribution`.

<br>

__5. competeMicrobiota__

The last function is to show potential microbial competition for various sugars. Each corner is a potential sugar degradation pathway, the the score is the relative completeness of this pathway, calculated as available enzymes divided by total enzymatic steps (different versions of the same enzyme are counted as the same for this function).

Run the code, play with the interactive plot, and compare your favourite ones.

```{r compMicro, message=FALSE, warning=FALSE}
library("microCompet")
genomeName <- "L. johnsonii"
ER <- microCompet::EnzymaticReactions
ED <- microCompet::EnzymeDistribution
fullEnzymeGeneList <- ED$Gene
# No need to rerun extractCarboGenes (lines with #) if you ran the previous example
genomeFilePath <- system.file("extdata",                              #
                              "Lactobacillus_johnsonii.gb",           #
                              package = "microCompet",                #
                              mustWork = TRUE)                        #
carboGenes <- extractCarboGenes(genomeFilePath, fullEnzymeGeneList)   #
firstMicrobe <- 5
lastMicrobe <- 13
compMicro <- competeMicrobiota(genomeName, carboGenes, ER,
                               ED, firstMicrobe, lastMicrobe)
compMicro
```


## Yeah!

Now you understand how to play with this package, have fun!


## References

1. Butts, C. 2008. “network: a Package for Managing Relational Data in R.” _Journal of Statistical Software_. 24(2). <URL: https://www.jstatsoft.org/v24/i02/paper>.

2. Butts, C. 2020. network: Classes for Relational Data. The Statnet Project (<URL:http://www.statnet.org>). R package version 1.16.1, <URL: https://CRAN.R-project.org/package=network>.

3. Butts, C.T. 2020. sna: Tools for Social Network Analysis. R package version 2.6. https://CRAN.R-project.org/package=sna

4. Csardi, G., Nepusz, T. 2006. The igraph software package for complex network research, InterJournal, Complex Systems 1695. https://igraph.org

5. Grolemund, G. (2015). Learn Shiny - Video Tutorials. https://shiny.rstudio.com/tutorial/

6. Karp, P.D., Riley, M., Paley, S.M., and Pellegrini-Toole A. 2002. The MetaCyc Database. _Nucleic Acids Re_. 30(1):59-61. doi:10.1093/nar/30.1.59

7. National Center for Biotechnology Information (NCBI)[Internet]. Bethesda (MD): National Library of Medicine (US), National Center for Biotechnology Information [1988]. Available from: https://www.ncbi.nlm.nih.gov/

8. Pedersen, T.L. 2020. ggraph: An Implementation of Grammar of Graphics for Graphs and Networks. R package version 2.0.3. https://CRAN.R-project.org/package=ggraph

9. R Core Team. 2020. R: A language and environment for statistical computing. R Foundation for Statistical Computing. Vienna, Austria. URL: https://www.R-project.org/.

10. Wickham, H. 2016. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York.

11. Wickham, H. 2020. R Packages. https://r-pkgs.org/

12. Xie, Y. 2020. bookdown: Authoring Books and Technical Documents with R Markdown. https://bookdown.org/yihui/bookdown/

## Session Info

This list seems longer than expected, but that's the case #_#
```{r}
sessionInfo()
```
